### PR TITLE
Cache the most recent Seed/Circulation calls to reduce pressure on ledger

### PR DESCRIPTION
**Pull request by tsachiherman**
## Summary

This PR adds a cache for the data.Ledger for the `Circulation` and `Seed` calls.
The goal of this PR is to reduce the contingency pressure on the downstream accounts update lock.

The high-level plan is to eventually eliminate this code, and make sure that the agreement service is sufficiently smart to perform a similar logic.

## Test Plan

Unit test added.

### CODE REVIEW

File: data/ledger.go [hunk 1]
```diff
@@ -43,6 +44,23 @@ type Ledger struct {
...
+
+type roundCirculation struct {
+	prevRound       basics.Round
+	prevOnlineMoney basics.MicroAlgos
+	curRound        basics.Round
+	curOnlineMoney  basics.MicroAlgos
+}
+
+type roundSeed struct {
+	prevRound basics.Round
```
Comment (nicholasguoalgorand) [inline data/ledger.go:60]: prevRound indicates the previously stored round, not the most previous round right? Might be helpful to include a comment clarifying this / rename the field

**Comment by nicholasguoalgorand** [comment]
Do we know if this makes a performance improvement or are we just guessing that it does? Are there any specific figures that would be helpful to include in the pr?

File: data/ledger.go (hunk 1 above)
Comment (tsachiherman) [reply data/ledger.go:60]: I'm not sure what `most previous round` would be. It's a two element moving window cache.
The implementation here is the trivial one, allowing to reduce the pressure from the accounts update.
My intent is eventually to wrap this into the agreement itself, and delete this code completely.
( when the agreement is making ledger advancements, it should query these values and store these locally.
we should go "outside" the agreement "box" only when needed )

**Comment by tsachiherman** [comment]
> Do we know if this makes a performance improvement or are we just guessing that it does? Are there any specific figures that would be helpful to include in the pr?

I have measured that it reduces the timing from 800-1000ns to 350-450ns. However, I'm less interested in the absolute gains here, and more concern about lowering the pressure on the ledger code itself. See my previous nodes about moving further.

File: data/ledger.go (hunk 1 above)
Comment (nicholasguoalgorand) [reply data/ledger.go:60]: right so `prevRound` could be confused with `curRound - 1`. But this is not the case here, so I suggested a comment possibly clarifying.

Comment (tsachiherman) [reply data/ledger.go:69]: I've added some documentation which I hope would be clear enough.

Comment (tsachiherman) [reply data/ledger.go:69]: I'm not a big fan of a/b; how do you feel about an array of two elements ?

Comment (brianolson) [reply data/ledger.go:69]: 2 element array sounds great

Comment (tsachiherman) [reply data/ledger.go]: ok, done.

**Review by brianolson — APPROVED** [review]

**Pull request merged**